bpo-40452

E-Paine opened the pull request:
<!-- issue-number: [bpo-40452](https://bugs.python.org/issue40452) -->
https://bugs.python.org/issue40452
<!-- /issue-number -->

E-Paine committed:
Windows keep clipboard on close

On Windows, the tkinter clipboard gets cleared with the closure of the root. This commit forces IDLE (on Windows only) to preserve the clipboard contents even after all windows of an instance have been closed.

E-Paine committed:
Imported required libraries

Imported the libraries required for the previous commit

E-Paine committed:
Only call when the root is actually closing

Before, the call would occur whenever a toplevel closed, regardless of whether the root was also closing (which has now been fixed)

ZackerySpytz commented on Lib/idlelib/filelist.py:
```diff
@@ -66,6 +69,18 @@ def unregister_maybe_terminate(self, edit):
+            try:
+                clip = self.root.clipboard_get().encode('utf-8')
+                if clip and sys.platform[:3] == 'win' and \
+                   len(clip) <= io.DEFAULT_BUFFER_SIZE:  # Required to avoid exceeding the pipe bufsize
```
Please limits lines to 79 chars.

terryjreedy replied:
Fixed, but next line was not.  I am fixing latter.

ZackerySpytz commented on Lib/idlelib/filelist.py:
```diff
@@ -66,6 +69,18 @@ def unregister_maybe_terminate(self, edit):
+                clip = self.root.clipboard_get().encode('utf-8')
+                if clip and sys.platform[:3] == 'win' and \
+                   len(clip) <= io.DEFAULT_BUFFER_SIZE:  # Required to avoid exceeding the pipe bufsize
+                    p = subprocess.Popen('clip', shell=True, stdin=subprocess.PIPE)
```
I think a `with` statement should be used here.

ZackerySpytz commented on Lib/idlelib/filelist.py:
```diff
@@ -66,6 +69,18 @@ def unregister_maybe_terminate(self, edit):
+                    p = subprocess.Popen('clip', shell=True, stdin=subprocess.PIPE)
+                    p.stdin.write(clip)
+                    p.stdin.close()
+            except:
```
Bare `except`s should not be used (https://bugs.python.org/issue15313 is for removing bare `except`s in IDLE).

terryjreedy replied:
Fixed.

ZackerySpytz commented on Lib/idlelib/filelist.py:
```diff
@@ -66,6 +69,18 @@ def unregister_maybe_terminate(self, edit):
+                    p.stdin.write(clip)
+                    p.stdin.close()
+            except:
+                # Ignore anything that goes wrong
```
I don't think three lines of comments are needed here.

E-Paine committed:
Addressed ZackerySpytz's comments

I have addressed the issues ZackerySpytz has raised in his comments (namely reducing line-length, using a with statement and changing a broad except). The other minor changes are:
 - Clarification has been added on why we work directly with stdin rather than the Popen class
 - Root quit method has been moved to a 'finally'
Please leave a comment if errors other than 'OSError's should be ignored

E-Paine committed:
Catches empty clipboard errors

When the clipboard was empty, tkinter would raise a TclError (which wasn't being caught). My point still stands about commenting if there are errors which I have missed and should be ignored.

terryjreedy commented on Lib/idlelib/filelist.py:
```diff
@@ -66,7 +70,22 @@ def unregister_maybe_terminate(self, edit):
+                    with subprocess.Popen('clip', shell=True, stdin=subprocess.PIPE).stdin as p:
+                        p.write(clip)
+            except (OSError,
+                    _tkinter.TclError):
```
```suggestion
            except (OSError, tkinter.TclError):
```

terryjreedy commented on Lib/idlelib/filelist.py:
```diff
@@ -1,6 +1,10 @@
+import subprocess
+import sys
+import _tkinter
 from tkinter import messagebox as tkMessageBox
```
```suggestion
from tkinter import TclError, messagebox as tkMessageBox
```

terryjreedy commented on Lib/idlelib/filelist.py:
```diff
@@ -66,7 +70,22 @@ def unregister_maybe_terminate(self, edit):
+                clip = self.root.clipboard_get().encode('utf-8')
+                if clip and sys.platform[:3] == 'win' and \
+                   len(clip) <= io.DEFAULT_BUFFER_SIZE:
+                    # content size check required to avoid exceeding the pipe bufsize
```
```suggestion
                    # Avoid exceeding the pipe bufsize.
```

terryjreedy replied:
done

terryjreedy commented on Lib/idlelib/filelist.py:
```diff
@@ -66,7 +70,22 @@ def unregister_maybe_terminate(self, edit):
+                   len(clip) <= io.DEFAULT_BUFFER_SIZE:
+                    # content size check required to avoid exceeding the pipe bufsize
+                    
+                    # we work directly with stdin as '.communicate' & '.__exit__' call '.wait'
```
```suggestion
                    
                    # '.communicate' & '.__exit__' call '.wait'.
```
Put this after p.write.

terryjreedy replied:
fixed

terryjreedy commented on Lib/idlelib/filelist.py:
```diff
@@ -66,7 +70,22 @@ def unregister_maybe_terminate(self, edit):
+            except (OSError,
+                    _tkinter.TclError):
+                # 'OSError's can probably be ignored (others may not)
+                # 'TclError's are due to an empty clipboard
```
```suggestion
                # 'TclError's are due to an empty clipboard.
```
Delete line, make other a proper sentence, as recommended in PEP 8.

terryjreedy replied:
Done.

terryjreedy reviewed (approved):
See comment about manual test on issue.
Remove trailing whitespace that caused Travis failure, and recheck with IDLE or patchcheck after other edits.

E-Paine committed:
committed terryjreedy's suggestions

I have committed @terryjreedy's suggested changes almost exactly, only changing '_tkinter.TclError' to 'TclError' for the new imports

E-Paine committed:
Changed encoding from UTF-8 to UTF-16

Changed encoding type from UTF-8 to UTF-16 (see comment for explanation)

terryjreedy committed:
Merge remote-tracking branch 'origin/master' into pr_19819

terryjreedy committed:
Wrap long line.

terryjreedy commented on Lib/idlelib/filelist.py:
```diff
@@ -66,7 +69,20 @@ def unregister_maybe_terminate(self, edit):
+                if clip and sys.platform[:3] == 'win' and \
+                   len(clip) <= io.DEFAULT_BUFFER_SIZE:
+                    # Avoid exceeding the pipe bufsize.
+                    with subprocess.Popen('clip', shell=True,
```
`shell=True` is not needed, I am deleting it.

E-Paine replied:
I added it because, without it I found that a console window flashes temporarily (this may just be because I am on a slower computer and there also may be a better way of resolving it).

terryjreedy replied:
I saw flashes with one of my other experiments, but never with this.  A mute point for the moment.  Anything we do should be tried on multiple machines.

terryjreedy commented on Lib/idlelib/filelist.py:
```diff
@@ -66,7 +69,20 @@ def unregister_maybe_terminate(self, edit):
+                                          stdin=subprocess.PIPE).stdin as p:
+                        p.write(clip)
+                        # Popen '.communicate' & '.__exit__' call '.wait'.
+            except (OSError, TclError):
```
I don't believe OSError is possible. In any case, we should try to write code that will not fail under anything like normal circumstances.  Deleting this.

terryjreedy committed:
Rearrange code.

terryjreedy reviewed (changes requested):
I rearranged a bit.  The patch fixes the paste-after close issue, but while exit is immediate (as signified by the Command Prompt > prompt), the patch is buggy in that it leaves the subprocess running.  This is visible when running from a console.
```
f:\dev\3x>python -m idlelib
Running Debug|Win32 interpreter...

Warning (from warnings module):
  File "f:\dev\3x\lib\subprocess.py", line 1048
    _warn("subprocess %s is still running" % self.pid,
ResourceWarning: subprocess 12596 is still running
>>>
f:\dev\3x>
```
I tried sending Windows EOF ^s\n as bytes, but no effect.  Ditto for adding p.close().
I tried subprocess.run('clip', input=clip) and complete exit is delayed for about 4 seconds instead of printing the warning.  Neither is acceptable.
Pasting into IDLE before closing 'solves' the problem.  Instead of using clip, I tried simulating this with a Toplevel and Text but so far failed.
So a change is needed but I don't know what.

Clearing the clipboard when IDLE closes strikes me as a bug somewhere.  If not in IDLE's closing, it should be possible to write a short tkinter app showing the same issue when run directly, without IDLE.
@serhiy-storchaka Do you have any ideas as to what might be going on or how to better fix?

clip is meant to be used with in Command Prompt to put program output, either to stdout or a file, on the clipboard.
@eryksun Do you know what Windows system call clip uses?  Can we access it via ctypes (as done near the top of pyshell for resolution settings)?  Or do you know any other alternative to the current patch?
